drivers/imx-enet.c

```diff
@@ -465,121 +592,165 @@ static void enet_irq_thread(void *arg)
+		hold = ENET_MSCR_HOLDTIME >> ENET_MSCR_HOLDTIME_SHIFT;
 
+	/* if hold time is too low - set it to min */
 	if (hold < 0)
 		hold = 0;
-	else if (hold > (ENET_MSCR_HOLDTIME_MASK >> ENET_MSCR_HOLDTIME_SHIFT))
-		hold = ENET_MSCR_HOLDTIME_MASK;
-	else
-		hold <<= ENET_MSCR_HOLDTIME_SHIFT;
-
-	if (hold > (state->mscr & ENET_MSCR_HOLDTIME_MASK)) {
-		state->mscr &= ~ENET_MSCR_HOLDTIME_MASK;
-		state->mscr |= hold;
+	/* if hold time is too large - set it to max */
+	if (hold > (ENET_MSCR_HOLDTIME >> ENET_MSCR_HOLDTIME_SHIFT))
+		hold = ENET_MSCR_HOLDTIME;
+
+	/* if hold time is larger than current, set the value in the reg */
+	if (hold > (state->mmio->MSCR & ENET_MSCR_HOLDTIME)) {
+		state->mmio->MSCR &= ~ENET_MSCR_HOLDTIME;
+		state->mmio->MSCR |= hold << ENET_MSCR_HOLDTIME_SHIFT;
+#if ENET_DEBUG
 		changed = 1;
+#endif
 	}
 
-	if (!opt_preamble && (state->mscr & ENET_MSCR_DIS_PRE)) {
-		state->mscr &= ~ENET_MSCR_DIS_PRE;
+	/* if the preamble should be disabled but isn't: */
+	if (!opt_preamble && (state->mmio->MSCR & ENET_MSCR_DIS_PRE)) {
+		state->mmio->MSCR &= ~ENET_MSCR_DIS_PRE;
+#if ENET_DEBUG
 		changed = 1;
+#endif
 	}
 
+#if ENET_DEBUG
 	if (changed) {
-		speed = (state->mscr & ENET_MSCR_SPEED_MASK) >> ENET_MSCR_SPEED_SHIFT;
-		hold = (state->mscr & ENET_MSCR_HOLDTIME_MASK) >> ENET_MSCR_HOLDTIME_SHIFT;
-		enet_printf(state, "mdio: speed %u (%u kHz), hold %u (%u ns), %s preamble",
-				speed, ENET_CLK_KHZ / 2 / (speed + 1),
-				hold, (hold + 1) * 1000000 / ENET_CLK_KHZ,
-				state->mscr & ENET_MSCR_DIS_PRE ? "no" : "with");
-	}
+		speed = (state->mmio->MSCR & ENET_MSCR_MII_SPEED) >> ENET_MSCR_MII_SPEED_SHIFT;
+		hold = (state->mmio->MSCR & ENET_MSCR_HOLDTIME) >> ENET_MSCR_HOLDTIME_SHIFT;
 
-	if (MDC_ALWAYS_ON)
-		state->mmio->MSCR = state->mscr;
+		enet_printf(state, "mdio setup: speed %u (%u kHz), hold %u (%u ns), %s preamble",
+			speed, ENET_CLK_KHZ / 2 / (speed + 1),
+			hold, (hold + 1) * 1000000 / ENET_CLK_KHZ,
+			state->mmio->MSCR & ENET_MSCR_DIS_PRE ? "no" : "with");
+	}
+#endif
 
 	return 0;
 }
 
 
-static void enet_mdioWait(enet_priv_t *state)
+static void enet_mdioWait(enet_state_t *state, time_t timeout_us)
 {
-	// FIXME: timeout
-	while (!(state->mmio->EIR & ENET_IRQ_MII))
-		/* relax */;
-	state->mmio->EIR = ENET_IRQ_MII;
+	time_t now, when;
+	gettime(&now, NULL);
+	when = now + timeout_us * 1000;
+
+	while (!(state->mmio->EIR & ENET_IRQ_MII)) {
+		if (timeout_us) {
+			gettime(&now, NULL);
+			if (now >= when) {
+#if MDIO_DEBUG
+				enet_printf(state, "enet_mdioWait: WARN: timeout");
+#endif
+				break;
+			}
+		}
+	}
+
+	state->mmio->EIR |= ENET_IRQ_MII;
+	enet_dataSyncBarrier();
 }
 
 
-static uint16_t enet_mdioIO(enet_priv_t *state, unsigned addr, unsigned reg, unsigned val, int read)
+static uint16_t enet_mdioIO(enet_state_t *state, unsigned addr, unsigned reg, uint16_t val, unsigned op)
 {
-	state->mmio->EIR = ENET_IRQ_MII;
-	if (!MDC_ALWAYS_ON)
-		state->mmio->MSCR = state->mscr;
+#if MDIO_DEBUG
```

Comment:
not sure about `MDC_ALWAYS_ON` functionality - but if was previously enabled and now is missing